Sources_v3/Controllers/CurrentUserController_Tests.swift

```diff
@@ -0,0 +1,355 @@
+    
+    // MARK: - Delegate
+    
+    func test_delegate_isAssignedCorrectly() {
+        let delegate = TestDelegate()
+        
+        // Set the delegate
+        controller.delegate = delegate
+        
+        // Assert the delegate is assigned correctly
+        XCTAssert(controller.delegate === delegate)
+    }
+
+    func test_delegate_isNotifiedAboutStateChanges() throws {
+        // Set the delegate
+        let delegate = TestDelegate()
+        delegate.expectedQueueId = controllerCallbackQueueID
+        controller.delegate = delegate
+        
+        // Assert no state changes received so far
+        XCTAssertNil(delegate.state)
+        
+        // Start updating
+        let startUpdatingError = try await(controller.startUpdating)
+        
+        // Assert `startUpdating` finished without any error
+        XCTAssertNil(startUpdatingError)
+        
+        // Assert delegate is notified about state changes
+        AssertAsync.willBeEqual(delegate.state, .localDataFetched)
+    }
+    
+    func test_genericDelegate_isNotifiedAboutStateChanges() throws {
+        // Set the delegate
+        let delegate = TestDelegateGeneric()
+        delegate.expectedQueueId = controllerCallbackQueueID
+        controller.setDelegate(delegate)
+        
+        // Assert no state changes received so far
+        XCTAssertNil(delegate.state)
+        
+        // Start updating
+        let startUpdatingError = try await(controller.startUpdating)
+        
+        // Assert `startUpdating` finished without any error
+        XCTAssertNil(startUpdatingError)
+        
+        // Assert delegate is notified about state changes
+        AssertAsync.willBeEqual(delegate.state, .localDataFetched)
+    }
+    
+    func test_delegate_isNotifiedAboutCreatedUser() throws {
+        let extraData = NameAndImageExtraData(name: .unique, imageURL: .unique())
+        let currentUserPayload: CurrentUserPayload<DefaultDataTypes.User> = .dummy(userId: .unique,
+                                                                                   role: .user,
+                                                                                   extraData: extraData)
+        
+        // Set the delegate
+        let delegate = TestDelegate()
+        delegate.expectedQueueId = controllerCallbackQueueID
+        controller.delegate = delegate
+
+        // Start updating
+        let startUpdatingError = try await(controller.startUpdating)
+        
+        // Assert `startUpdating` finished without any error
+        XCTAssertNil(startUpdatingError)
+
+        // Simulate saving current user to a database
+        let createUserError = try await {
+            client.databaseContainer.write({ session in
+                try session.saveCurrentUser(payload: currentUserPayload)
+            }, completion: $0)
+        }
+        
+        // Assert saving current user to a database finished without any error
+        XCTAssertNil(createUserError)
```

Comment:
☝️

Reply:
fixed in 9b3c74ee2d704845b8cd83448ade208720df8d94